libraries/net/node.cpp

```diff
@@ -283,6 +193,97 @@ namespace graphene { namespace net { namespace detail {
 #define MAXIMUM_NUMBER_OF_BLOCKS_TO_HANDLE_AT_ONE_TIME 200
 #define MAXIMUM_NUMBER_OF_BLOCKS_TO_PREFETCH (10 * MAXIMUM_NUMBER_OF_BLOCKS_TO_HANDLE_AT_ONE_TIME)
 
+   /******
+    * Use information passed from command line or config file to advertise nodes
+    */
+   class list_address_builder : public node_impl::address_builder
+   {
+      public:
+      list_address_builder(fc::optional<std::vector<std::string>> address_list)
+      {
+         if (address_list.valid())
+         {
+            advertise_list.reserve( address_list->size() );
+            auto& list = advertise_list;
+            std::for_each( address_list->begin(), address_list->end(), [&list]( std::string str ) {
+                  // ignore fc exceptions (like poorly formatted endpoints)
+                  try
+                  {
+                     list.emplace_back( graphene::net::address_info(
+                        fc::ip::endpoint::from_string(str),
+                        fc::time_point_sec(),
+                        fc::microseconds(0),
+                        node_id_t(),
+                        peer_connection_direction::unknown,
+                        firewalled_state::unknown ));
+                  }
+                  catch(const fc::exception& ) {
+                     wlog( "Address ${addr} invalid.", ("addr", str) );
+                  } 
+               } );
+         }
+      }
+
+      void build(node_impl* impl, address_message& reply)
+      {
+         reply.addresses = advertise_list;
+      }
+      private:
+      std::vector<graphene::net::address_info> advertise_list;
+   };
+
+   /****
+    * Advertise all nodes except a predefined list
+    */
+   class exclude_address_builder : public node_impl::address_builder
+   {
+      public:
+      exclude_address_builder(const fc::optional<std::vector<std::string>>& address_list)
+      {
+         if (address_list.valid())
+         {
+            std::for_each(address_list->begin(), address_list->end(), [&exclude_list = exclude_list](std::string input)
+                {
+                  exclude_list.insert(input);
+                });
+         }
+      }
+      void build(node_impl* impl, address_message& reply)
+      {
+         reply.addresses.reserve(impl->_active_connections.size());
+         for(const peer_connection_ptr& active_peer : impl->_active_connections)
+         {
+            if (!exists_in_list(active_peer, exclude_list))
+               reply.addresses.emplace_back(update_address_record(impl, active_peer));
+         }
+         reply.addresses.shrink_to_fit();
+      }
+      private:
+      fc::flat_set<std::string> exclude_list;
+      bool exists_in_list(const peer_connection_ptr& peer, const fc::flat_set<std::string>& exclude_list) const
+      {
+         std::string remote = *peer->get_remote_endpoint();
+         if ( std::find(exclude_list.begin(), exclude_list.end(), remote) != exclude_list.end() )
```

Comment:
`exclude_list.find(remote)` is more efficient because it uses binary seach not linear.

Also can return the comparison result directly instead of `if (expr) return true; else return false;`.